archive/indexer.go

```diff
@@ -0,0 +1,113 @@
+package archive
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/brimsec/zq/pkg/sst"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zio/detector"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+const zarExt = ".zar"
+
+// TBD
+type Indexer interface {
+	Create(*sst.Writer, zbuf.Reader)
+	//Search([]byte) bool
+}
+
+//XXX this is a test stub that creates simple indexes of IP addresses
+func CreateIndexes(dir string) error {
+	nerr := 0
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("%q: %v", path, err)
+		}
+		name := info.Name()
+		if info.IsDir() {
+			if filepath.Ext(name) == zarExt {
+				//XXX need to merge into or replace existing index
+				return filepath.SkipDir
+			}
+			// descend...
+			return nil
+		}
+		if filepath.Ext(name) == ".bzng" {
+			err = IndexLogFile(path)
+			if err != nil {
+				fmt.Printf("%s: %s\n", path, err)
+				nerr++
+				if nerr > 10 {
+					//XXX
+					return errors.New("stopping after too many errors...")
+				}
+			}
+			// drop through and continue
+		}
+		return nil
+	})
+	return err
+}
+
+func IndexLogFile(path string) error {
+	subdir := path + zarExt
+	sstName := "sst:type:ip"
+	sstPath := filepath.Join(subdir, sstName)
+	// XXX remove without warning, should have force flag
+	sst.Remove(sstPath)
+
+	fmt.Printf("%s: indexing as %s\n", path, sstPath)
+
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	reader, err := detector.LookupReader("bzng", file, resolver.NewContext())
+	if err != nil {
+		return err
+	}
+	table, err := indexTypeIP(reader)
+	if err != nil {
+		return err
+	}
+	if table.Size() == 0 {
+		//XXX
+		return errors.New("nothing to index")
```

Comment:
Yes, this would be a better approach but the current logic doesn't handle the empty table.  I would prefer to handle this later when we soon switch over to zng.